### PR TITLE
Launcher: explicitly handle cli arguments to be passed to the Component

### PR DESCRIPTION
**Pull request by qwint**
## What is this fixing or adding?
adds handling for the -- cli arg by having launcher capture, ignore, and pass through all of the values after it, while only processing (and validating) the values before it

updates text client and its components to allow for args to be passed through, captured in run_as_textclient, and used in parse_args if present

updates help text (both the usage line if parsing fails and --help text) to explicitly mention the `--` arg parsing (see below)

## How was this tested?
below errors as well as confirming the following work as expected
`python Launcher.py "Text Client" -- --connect test@:localhost`
`python CommonClient.py --connect test@:localhost`

I think url handling would need to be overhauled, but #2779 is already trying to do that, but currently the following work
`python Launcher.py -- archipelago://test:@localhost`
`python CommonClient.py archipelago://test:@localhost`
and opening up `archipelago://test:@localhost` in browser and letting Archipelago handle the protocol 

## If this makes graphical changes, please attach screenshots.
`python Launcher.py test test`
image

`python Launcher.py "Text Client" --nogui`
image

`python Launcher.py "Text Client" -- test test`
image

`python Launcher.py --help`
image

`python CommonClient.py --help`
image

`python CommonClilent.py test test`
image

### CODE REVIEW

File: worlds/LauncherComponents.py [hunk 1]
```diff
@@ -56,10 +56,10 @@ def __repr__(self):
 processes = weakref.WeakSet()
 
 
-def launch_subprocess(func: Callable, name: str = None):
+def launch_subprocess(func: Callable, name: str = None, args=()):
```
Comment (alwaysintreble) [inline worlds/LauncherComponents.py:59]:
```suggestion
def launch_subprocess(func: Callable, name: str = None, args: Tuple[str, ...] = ()) -> None:
```

**Review by alwaysintreble — APPROVED** [review]
some minor nitpicks. tested it and it works and error message is clearer about the intended usage of passthrough args

File: CommonClient.py [hunk 2]
```diff
@@ -1033,7 +1033,7 @@ async def main(args):
     parser = get_base_parser(description="Gameless Archipelago Client, for text interfacing.")
     parser.add_argument('--name', default=None, help="Slot Name to connect as.")
     parser.add_argument("url", nargs="?", help="Archipelago connection url")
-    args = parser.parse_args()
+    args = parser.parse_args(args if args else None)
```
[the remark this answers is not in the document]
Comment (qwint) [reply CommonClient.py:1036]: it didn't work when i tried but i can retest

[the remark this answers is not in the document]
Comment (qwint) [reply CommonClient.py:1036]: yea, dropping to just parse_args(args) does not work when using `python CommonClient.py args`
potentially we can pass something into run_as_textclient() but the things i tested all didn't pass None into parse_args() correctly

**Review by remyjette — APPROVED** [review]
Tested this with my changes to BizHawkClient in #3554 and it works as expected, tried both no args, with a patch file, with positional args like `--help` or `--connect` and it all worked correctly

**Review by nicholassaylor — APPROVED** [review]
`python Launcher.py -- archipelago://test:@localhost` working seems a bit odd, but it ultimately outside of the scope of this PR

All of the examples worked on my system as well, and a few odd combinations I could try and come up with returned errors similar to what I would expect.

Did not look at code too closely, but the client-facing changes are appropriate and the passthrough args make sense.

Look at the PR as a whole, I couldn't find anything of concern that wasn't covered in code already besides the url concern that is covered in the PR mentioned by qwint.

File: Launcher.py [hunk 3]
```diff
@@ -322,15 +322,28 @@ def main(args: Optional[Union[argparse.Namespace, dict]] = None):
...
+                           dest="args", default=[])
+    args = sys.argv[1:]
+    if "--" in args:
+        i = args.index("--")
+        passthrough_args = args[i+1:]
+        args = args[:i]
+        args = parser.parse_args(args)
+        args.args = passthrough_args
+    else:
+        args = parser.parse_args()
```
Comment (NewSoupVi) [inline Launcher.py:345]: Is this part necessary? `parser.parse_args` already seems to understand how to handle `--` correctly

File: worlds/LauncherComponents.py [hunk 4]
```diff
@@ -78,9 +78,9 @@ def __call__(self, path: str) -> bool:
         return False
 
 
-def launch_textclient():
+def launch_textclient(*args):
```
Comment (NewSoupVi) [inline worlds/LauncherComponents.py:81]: In Launcher.py `main()`, do you think we could change

`run_component(args["component"], *args["args"])`

to

```
try:
    run_component(args["component"], *args["args"])
except TypeError as e:
    warn(f"Component {args["component"]} does not know how to take args {args["args"]}."}
    run_component(args["component"])
```

Comment (NewSoupVi) [reply worlds/LauncherComponents.py:81]: Idk if `except TypeError as e` is the best way to check for "this function does not take this many arguments", if there is a better way that would be preferable ofc

**Comment by NewSoupVi** [comment]
I have two comments I'd like to be answered, otherwise this is more or less approved from my side

File: worlds/LauncherComponents.py (hunk 4 above)
Comment (qwint) [reply worlds/LauncherComponents.py:81]: `if len(inspect.signature(x).parameters)` looks to be the most direct way if you really want to check there are any args defined
are you just asking for it so this is less of a breaking change (when args are used, which isn't very often) or do you actually think adding some code to strip args if unneeded is important to the ecosystem?

File: Launcher.py (hunk 3 above)
Comment (qwint) [reply Launcher.py:345]: I think i figured out what the issue with this was (after reverting the `--` arg in the Parser), it joins args before and after the `--` arg together, so this happens
```
> python launcher.py test1, test2 -- test3 --test4
Namespace(update_settings=False, args=['test2', 'test3', '--test4'], **{'Patch|Game|Component': 'test1,'})
```

this is not a hill i'm willing to die on though, 

tested the rollback on these lines (and the 'args' arg) in source and frozen and the only scenarios that didn't work as expected were negative tests that just had invalid args ignored ex. `python launcher.py test test`

will commit if you confirm this is good enough:tm: and a worthy downside to get rid of the manual parsing

File: worlds/LauncherComponents.py (hunk 4 above)
Comment (NewSoupVi) [reply worlds/LauncherComponents.py:81]: Yeah idk this is probably not that important you're right

File: Launcher.py (hunk 3 above)
Comment (qwint) [reply Launcher.py]: got the confirmation in discord

**Review by NewSoupVi — APPROVED** [review]

**Pull request merged**